gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/decoration/SideEffectStrategy.java

```diff
@@ -60,7 +61,11 @@ public static <A> void addSideEffect(final TraversalStrategies traversalStrategi
             strategy = cloneStrategy;
             traversalStrategies.addStrategies(strategy);
         }
-        strategy.sideEffects.add(new Triplet<>(key, value instanceof Supplier ? (Supplier) value : new ConstantSupplier<>(value), reducer));
+
+        // don't want the GValue to leak beyond strategy application or else the Supplier will start producing it
+        // during execution
+        final ConstantSupplier initialValue = value instanceof GValue ? new ConstantSupplier<>(((GValue) value).get()) : new ConstantSupplier<>(value);
+        strategy.sideEffects.add(new Triplet<>(key, value instanceof Supplier ? (Supplier) value : initialValue, reducer));
```

Comment:
can `value` be of `Supplier<GValue>` type?

Reply:
Yes that is possible. Nice catch! I've updated the logic such that any GValue supplied by a `Supplier<GValue>` will be unboxed as well.